File: src/src_platform_jfx/fmg/jfx/img/MosaicGroupImg.java
package fmg.jfx.img;

import java.util.stream.Collectors;
import java.util.stream.Stream;

import fmg.common.Color;
import fmg.common.Pair;
import fmg.common.geom.PointDouble;
import fmg.core.img.MosaicGroupController;
import fmg.core.img.MosaicGroupModel;
import fmg.core.types.EMosaicGroup;

/**
 * Representable {@link fmg.core.types.EMosaicGroup} as image
 * <br>
 * JFX impl
 *
 * @param <TImage> JFX specific image: {@link javafx.scene.image.Image} or {@link javafx.scene.canvas.Canvas})
 **/
public abstract class MosaicGroupImg<TImage> extends MosaicSkillOrGroupView<TImage, MosaicGroupModel> {

   /** @param group - may be null. if Null - representable image of EMosaicGroup.class */
   protected MosaicGroupImg(EMosaicGroup group) {
      super(new MosaicGroupModel(group));
   }

   @Override
   protected Stream<Pair<Color, Stream<PointDouble>>> getCoords() { return getModel().getCoords(); }

   @Override
   public void close() {
      getModel().close();
      super.close();
   }

   /////////////////////////////////////////////////////////////////////////////////////////////////////
   //    custom implementations
   /////////////////////////////////////////////////////////////////////////////////////////////////////

   /** MosaicsGroup image view implementation over {@link javafx.scene.canvas.Canvas} */
   static class Canvas extends MosaicGroupImg<javafx.scene.canvas.Canvas> {

      private CanvasJfx canvas = new CanvasJfx(this);

      /** @param skill - may be null. if Null - representable image of EMosaicGroup.class */
      public Canvas(EMosaicGroup group) { super(group); }

      @Override
      protected javafx.scene.canvas.Canvas createImage() { return canvas.create(); }

      @Override
      protected void drawBody() { draw(canvas.getGraphics()); }

   }

   /** MosaicsGroup image view implementation over {@link javafx.scene.image.Image} */
   static class Image extends MosaicGroupImg<javafx.scene.image.Image> {

      private ImageJfx img = new ImageJfx(this);

      /** @param skill - may be null. if Null - representable image of EMosaicGroup.class */
      public Image(EMosaicGroup group) { super(group); }

      @Override
      protected javafx.scene.image.Image createImage() {
         img.createCanvas();
         return null; // img.createImage(); // fake empty image
      }

      @Override
      protected void drawBody() {
         draw(img.getGraphics());
         setImage(img.createImage()); // real image
      }

   }

   /** MosaicsGroup image controller implementation for {@link Canvas} */
   public static class ControllerCanvas extends MosaicGroupController<javafx.scene.canvas.Canvas, MosaicGroupImg.Canvas> {

      public ControllerCanvas(EMosaicGroup group) {
         super(group==null, new MosaicGroupImg.Canvas(group));
      }

      @Override
      public void close() {
         getView().close();
         super.close();
      }

   }

   /** MosaicsGroup image controller implementation for {@link Image} */
   public static class ControllerImage extends MosaicGroupController<javafx.scene.image.Image, MosaicGroupImg.Image> {

      public ControllerImage(EMosaicGroup group) {
         super(group==null, new MosaicGroupImg.Image(group));
      }

      @Override
      public void close() {
         getView().close();
         super.close();
      }

   }

   ////////////// TEST //////////////
   public static void main(String[] args) {
      TestDrawing.testApp(() ->
         Stream.concat(Stream.of((EMosaicGroup)null),
                       Stream.of(EMosaicGroup.values()))
               .map(e -> new Pair<>(new MosaicGroupImg.ControllerCanvas (e),
                                    new MosaicGroupImg.ControllerImage(e)))
               .flatMap(x -> Stream.of(x.first, x.second))
               .collect(Collectors.toList())
      );
   }
   //////////////////////////////////

}

File: src/src_platform_ava/fmg/ava/img/MosaicGroupImg.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Avalonia.Controls;
using Avalonia.Media;
using Avalonia.Media.Imaging;
using Avalonia.Rendering;
using fmg.common.geom;
using fmg.core.img;
using fmg.core.types;
using fmg.ava.utils;

namespace fmg.ava.img {

   /// <summary> Representable <see cref="EMosaicGroup"/> as image.
   /// <br/>
   /// Avalonia impl
   /// </summary>
   public static class MosaicGroupImg {

      /// <summary> Representable <see cref="EMosaicGroup"/> as image: common implementation part </summary>
      public abstract class View<TImage> : MosaicSkillOrGroupView<TImage, MosaicGroupModel>
         where TImage : class
      {

         /// <param name="group">may be null. if Null - representable image of typeof(EMosaicGroup)</param>
         protected View(EMosaicGroup? group)
            : base(new MosaicGroupModel(group))
         { }

         /// <summary> get paint information of drawing basic image model </summary>
         protected override IEnumerable<Tuple<fmg.common.Color, IEnumerable<PointDouble>>> Coords { get => Model.Coords; }

         protected override void Disposing() {
            Model.Dispose();
            base.Disposing();
         }

      }


      /////////////////////////////////////////////////////////////////////////////////////////////////////
      //    custom implementations
      /////////////////////////////////////////////////////////////////////////////////////////////////////

      /// <summary> MosaicsGroup image view implementation over <see cref="RenderTargetBitmap"/> </summary>
      public class RenderTargetBmp : View<RenderTargetBitmap> {

         private IVisualBrushRenderer _vbr;

         /// <param name="group">may be null. if Null - representable image of typeof(EMosaicGroup)</param>
         public RenderTargetBmp(EMosaicGroup? group, IControl ctrl)
            : base(group)
         {
            _vbr = new ImmediateRenderer(ctrl);
         }

         protected override RenderTargetBitmap CreateImage() {
            //var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            return new RenderTargetBitmap((int)Size.Width, (int)Size.Height);
         }

         protected override void DrawBody() {
            using (var dc = Image.CreateDrawingContext(_vbr)) {
               using (var dc1 = new DrawingContext(dc)) {
                  DrawBody(dc1);
               }
            }
         }

         protected override void Disposing() {
            (_vbr as IDisposable)?.Dispose();
            _vbr = null;
         }
      }

   //   /// <summary> Representable <see cref="EMosaicGroup"/> as image.
   //   /// <br/>
   //   /// Canvas impl
   //   /// </summary>
   //   public class CanvasImgSrc : CommonImpl<Canvas> {
   //
   //      /// <param name="group">may be null. if Null - representable image of typeof(EMosaicGroup)</param>
   //      public CanvasImgSrc(EMosaicGroup? group)
   //         : base(group)
   //      { }
   //
   //      protected override Canvas CreateImage() {
   //         return new Canvas {
   //            Width = Size.Width,
   //            Height = Size.Height
   //         };
   //      }
   //
   //      protected override void DrawBody() {
   //         DrawBody(Image, true);
   //      }
   //   }

      /** MosaicsGroup image controller implementation for {@link Canvas} */
      public class ControllerRenderTargetBmp : MosaicGroupController<RenderTargetBitmap, MosaicGroupImg.RenderTargetBmp> {

         public ControllerRenderTargetBmp(EMosaicGroup? group, IControl ctrl)
            : base(!group.HasValue, new MosaicGroupImg.RenderTargetBmp(group, ctrl))
         { }

         protected override void Disposing() {
            View.Dispose();
            base.Disposing();
         }

      }

   }

}
